File: apps/web/lib/permissions.ts
"use client";

import { useAuthBootstrap } from "./auth-session";
import type { Permission, UserRole } from "./types";

const rolePermissions: Record<UserRole, readonly Permission[]> = {
  admin: [
    "server.view", "server.create", "server.control", "server.configure", "server.delete",
    "backup.manage", "world.manage", "mod.manage", "player.manage", "share.manage",
    "node.manage", "team.manage", "settings.manage", "system.manage"
  ],
  member: [
    "server.view", "server.create", "server.control", "server.configure",
    "backup.manage", "world.manage", "mod.manage", "player.manage", "share.manage"
  ],
  viewer: ["server.view"]
};

export function permissionsForRole(role: UserRole): readonly Permission[] {
  return rolePermissions[role];
}

export function usePermissions() {
  const authQuery = useAuthBootstrap();

  const account = authQuery.data?.account;
  const role: UserRole = account?.role ?? (authQuery.data?.initialized === false ? "admin" : "viewer");
  const platformRole = account?.platformRole ?? (authQuery.data?.initialized === false ? "platform_admin" : "user");
  const permissions = new Set<Permission>(account?.permissions ?? permissionsForRole(role));
  const can = (permission: Permission) => permissions.has(permission);

  const isViewer = role === "viewer";
  const isMember = role === "member";
  const isAdmin = platformRole === "platform_admin";

  return {
    account,
    role,
    platformRole,
    permissions,
    can,
    isViewer,
    isMember,
    isAdmin,
    isLoading: authQuery.isLoading,
    // 细粒度权限判定
    canCreateServer: can("server.create"),
    canControlServer: can("server.control"),
    canEditServerConfig: can("server.configure"),
    canDeleteServer: can("server.delete"),
    canManageBackups: can("backup.manage"),
    canManageWorlds: can("world.manage"),
    canManageMods: can("mod.manage"),
    canManagePlayers: can("player.manage"),
    canManageShares: can("share.manage"),
    canManageNodes: can("node.manage"),
    canManageTeam: can("team.manage"),
    canEditSettings: can("settings.manage"),
    canManageSystem: can("system.manage"),
    canAccessGameAssets: !isViewer
  };
}

File: apps/api/internal/domain/permissions.go
package domain

type Permission string

const (
	PermissionServerView      Permission = "server.view"
	PermissionServerCreate    Permission = "server.create"
	PermissionServerControl   Permission = "server.control"
	PermissionServerConfigure Permission = "server.configure"
	PermissionServerDelete    Permission = "server.delete"
	PermissionBackupManage    Permission = "backup.manage"
	PermissionWorldManage     Permission = "world.manage"
	PermissionModManage       Permission = "mod.manage"
	PermissionPlayerManage    Permission = "player.manage"
	PermissionShareManage     Permission = "share.manage"
	PermissionNodeManage      Permission = "node.manage"
	PermissionTeamManage      Permission = "team.manage"
	PermissionSettingsManage  Permission = "settings.manage"
	PermissionSystemManage    Permission = "system.manage"
)

var viewerPermissions = []Permission{
	PermissionServerView,
}

var memberPermissions = []Permission{
	PermissionServerView,
	PermissionServerCreate,
	PermissionServerControl,
	PermissionServerConfigure,
	PermissionBackupManage,
	PermissionWorldManage,
	PermissionModManage,
	PermissionPlayerManage,
	PermissionShareManage,
}

var adminPermissions = []Permission{
	PermissionServerView,
	PermissionServerCreate,
	PermissionServerControl,
	PermissionServerConfigure,
	PermissionServerDelete,
	PermissionBackupManage,
	PermissionWorldManage,
	PermissionModManage,
	PermissionPlayerManage,
	PermissionShareManage,
	PermissionNodeManage,
	PermissionTeamManage,
	PermissionSettingsManage,
	PermissionSystemManage,
}

func NormalizeAccountRole(role Role) Role {
	if role == "" {
		return RoleAdmin
	}
	return role
}

func PermissionsForRole(role Role) []Permission {
	var permissions []Permission
	switch NormalizeAccountRole(role) {
	case RoleAdmin, RoleOwner:
		permissions = adminPermissions
	case RoleMember:
		permissions = memberPermissions
	default:
		permissions = viewerPermissions
	}
	return append([]Permission(nil), permissions...)
}

func RoleHasPermission(role Role, permission Permission) bool {
	for _, allowed := range PermissionsForRole(role) {
		if allowed == permission {
			return true
		}
	}
	return false
}
